File: AI_FSM/Source/AI_FSM/Private/FSM/FSMObject.cpp
// Fill out your copyright notice in the Description page of Project Settings.

#include "FSM/FSMComponent.h"
#include "FSM/FSMObject.h"


void UFSMObject::StartFSM(UFSMComponent* _owner)
{

	currentFSMComponent = _owner;
	desactivateMS = _owner->GetDesactivateMS();
	SetNextState(startingState);
}

void UFSMObject::SetNextState(TSubclassOf<UStateObject> _state)
{
	if(!_state)
		return;
	if (currentState)
		currentState->Exit();
	currentState = NewObject<UStateObject>(this, _state);
	currentState->Enter(this);
}

void UFSMObject::UpdateFSM()
{
	if (currentState)
		currentState->Update();
}

void UFSMObject::StopFSM()
{
	if (!currentState)
		return;
	currentState->Exit();
	currentState = nullptr;
}

File: AI_FSM/Assets/Scripte/FSM/FSMObject.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public class FSMObject: MonoBehaviour
{

    [SerializeField]
    private  State startingState = null;
    [SerializeField]
    private State currentState = null;
    [SerializeField, HideInInspector]
    FSMComponent component = null;
public FSMComponent Component =>component;
public void StartFSM(FSMComponent _owner)
{
        component = _owner;
        SetNextState(startingState);
    }
public void SetNextState(State _state)
    {
        if (_state==null)
            return;
        if (currentState !=null)
            currentState.Exit();
        currentState = _state;
        currentState.Enter(this);
    }
public virtual void UpdateFSM()
    {
        if (currentState == null)
            currentState.StateUpdate();
    }
public virtual void StopFSM()
    {
        if (currentState == null)
            return;
        currentState.Exit();
        currentState = null;
    }

}
